### PR TITLE
SL-123 credit card saving with hosted fields

### PR DESCRIPTION
**Pull request by tomjas1997**
With hosted fields there was no way to actually save credit cards, implemented it for PS 1.7 and PS 1.6.

### CODE REVIEW

File: src/Provider/PaymentRedirectionProvider.php [hunk 1]
```diff
@@ -59,67 +58,31 @@ public function __construct(Context $context, $moduleName, SaferPayFieldReposito
...
-        );
+        $paymentType = $this->paymentTypeProvider->get($paymentMethod);
 
-        if ($this->isIframeRedirect($paymentMethod)) {
-            $redirectionLink = $this->context->link->getModuleLink(
+        if ($paymentType === PaymentType::HOSTED_IFRAME) {
+            return $this->context->link->getModuleLink(
                 $this->moduleName,
-                'iframe',
+                'hostedIframe',
```
Comment (tomas862) [inline src/Provider/PaymentRedirectionProvider.php:66]: can be enums, no ? Or atleast constants

File: src/Provider/PaymentRedirectionProvider.php [hunk 2]
```diff
@@ -59,67 +58,31 @@ public function __construct(Context $context, $moduleName, SaferPayFieldReposito
...
             );
         }
 
-        if ($this->isHostedIframeRedirect($paymentMethod)) {
-            $redirectionLink = $this->context->link->getModuleLink(
+        if ($paymentType === PaymentType::IFRAME) {
+            return $this->context->link->getModuleLink(
                 $this->moduleName,
-                'hostedIframe',
+                'iframe',
```
Comment (tomas862) [inline src/Provider/PaymentRedirectionProvider.php:75]: same here

File: views/js/front/hosted-templates/hosted_fields.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        return;
+    }
+
+    $('[id="payment-form"]').on('submit', function (event) {
+        event.preventDefault();
+
+        var paymentType = $(this).find("[name=saferpayPaymentType]").val();
+
+        //NOTE: if it's not a hosted iframe then we don't need to submitHostedFields.
+        if (paymentType !== 'hosted_iframe') {
```
Comment (tomas862) [inline views/js/front/hosted-templates/hosted_fields.js:36]: same regarding enums, can be loaded on the page

File: views/js/front/hosted-templates/hosted_fields_16.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        return;
+    }
+
+    $('[id="payment-form"]').on('submit', function (event) {
+        event.preventDefault();
+
+        var paymentType = $(this).find("[name=saferpayPaymentType]").val();
+
+        //NOTE: if it's not a hosted iframe then we don't need to submitHostedFields.
+        if (paymentType !== 'hosted_iframe') {
```
Comment (tomas862) [inline views/js/front/hosted-templates/hosted_fields_16.js:36]: same

File: src/Provider/PaymentRedirectionProvider.php (hunk 1 above)
Comment (tomjas1997) [reply src/Provider/PaymentRedirectionProvider.php:66]: It's a  controleller name, not sure how we proceeded with that

Comment (tomjas1997) [reply src/Provider/PaymentRedirectionProvider.php:66]: Pretty sure we just had it as string, but i guess i could create an enum for controllers

Comment (tomas862) [reply src/Provider/PaymentRedirectionProvider.php:66]: yeah would be nice so newcommers would understand that it is somehow connected with the controller

**Review by tomas862 — APPROVED** [review]

File: controllers/front/ajax.php [hunk 5]
```diff
@@ -42,14 +43,23 @@ private function submitHostedFields()
...
             $initializeBody = $initializeService->initialize(
                 Tools::getValue('paymentMethod'),
                 (int) Tools::getValue(SaferPayConfig::IS_BUSINESS_LICENCE),
-                -1,
-                null,
-                Tools::getValue('fieldToken')
+                $selectedCard,
+                $alias,
+                Tools::getValue('fieldToken'),
+                'successHosted'
```
Comment (margud) [inline controllers/front/ajax.php:62]: i would prefer const here that could explain what this string means, but just option

**Pull request merged**